osmoutils/store_helper_test.go

```diff
@@ -56,6 +56,380 @@ func mockStop(b []byte) bool {
+			keyEnd:   []byte(prefixTwo + keyA),
+			parseFn: mockParseValue,
+
+			expectedErr: false,
+			expectedValues: []string{"0", "1"},
+		},
+		"different prefix, inserted out of lexicographic order": {
+			preSetKeys: []string{prefixOne + keyA, prefixTwo + keyA, prefixOne + keyB, prefixTwo + keyB},
+
+			keyStart: []byte(prefixOne + keyA),
+			keyEnd:   []byte(prefixTwo + keyA),
+			parseFn: mockParseValue,
+
+			expectedErr: false,
+			// should get all prefixOne keys as keys are stored in ascending lexicographic order
+			expectedValues: []string{"0", "2"},
+		},
+		"start key and end key same": {
+			preSetKeys: []string{prefixOne + keyA},
+
+			keyStart: []byte(prefixOne + keyA),
+			keyEnd:   []byte(prefixOne + keyA),
+			parseFn: mockParseValue,
+
+			expectedErr: false,
+			expectedValues: []string{},
+		},
+		"start key after end key": {
+			preSetKeys: []string{prefixOne + keyA, prefixOne + keyB, prefixOne + keyC},
+
+			keyStart: []byte(prefixOne + keyB),
+			keyEnd:   []byte(prefixOne + keyA),
+			parseFn: mockParseValue,
+
+			expectedErr: false,
+			expectedValues: []string{},
+		},
+		"get all values": {
+			preSetKeys: []string{prefixOne + keyA, prefixOne + keyB, prefixOne + keyC},
+
+			keyStart: nil,
+			keyEnd:   nil,
+			parseFn: mockParseValue,
+
+			expectedErr: false,
+			expectedValues: []string{"0", "1", "2"},
+		},
+		"get all values after start key": {
+			// SDK iterator is broken for nil end time, and non-nil start time
+			// https://github.com/cosmos/cosmos-sdk/issues/12661
+			// so we use []byte{0xff}
+			preSetKeys: []string{prefixOne + keyA, prefixOne + keyB, prefixOne + keyC},
+
+			keyStart: []byte(prefixOne + keyB),
+			keyEnd:   []byte{0xff},
+			parseFn: mockParseValue,
+
+			expectedErr: false,
+			expectedValues: []string{"1", "2"},
+		},
+		"parse with error": {
+			preSetKeys: []string{prefixOne + keyA, prefixOne + keyB, prefixOne + keyC},
+
+			keyStart: []byte(prefixOne + keyA),
+			keyEnd:   []byte(prefixOne + keyC),
+			parseFn: mockParseValueWithError,
+
+			expectedErr: true,
+		},
+	}
+
+	for name, tc := range testcases {
+		s.Run(name, func() {
+			s.SetupStoreWithBasePrefix()
+
+			for i, key := range tc.preSetKeys {
+				s.store.Set([]byte(key), []byte(fmt.Sprintf("%v", i)))
+			}
+
+			actualValues, err := osmoutils.GatherValuesFromStore(s.store, tc.keyStart, tc.keyEnd, tc.parseFn)
+
+			if tc.expectedErr {
+				s.Require().Error(err)
+				s.Require().Nil(actualValues)
+				return
+			}
+
+			s.Require().NoError(err)
+			s.Require().Equal(tc.expectedValues, actualValues)
+		})
+	}
+}
+
+func (s *TestSuite) TestGatherValuesFromStorePrefix() {
+
+	testcases := map[string]struct {
+		prefix     []byte
+		preSetKeys []string
+		parseFn    func(b []byte) (string, error)
+
+		expectedErr 	   bool
+		expectedValues []string
+	}{
+		"common prefix": {
+			preSetKeys: []string{prefixOne + keyA, prefixOne + keyB, prefixOne + keyC},
+			prefix: []byte(prefixOne),
+
+			parseFn: mockParseValue,
```

Comment:
Main comment and review on this PR: 
I think these test cases occur alot of times across different test cases, causing unnecessary duplication in test cases, what do you think about reducing these duplications by declaring these repetitive cases as variables or a separate method?

Reply:
Yeah I had originally decided against doing this since most test cases have minor but important differences specific to the function they're testing, but I see what you mean @mattverse. I just finished defining common key sets as variables – hopefully this addresses your point!